game/src/sandbox/dashboards/trip_table.rs

```diff
@@ -4,52 +4,177 @@ use abstutil::prettyprint_usize;
 use geom::{Duration, Time};
 use sim::{TripEndpoint, TripID, TripMode};
 use widgetry::table::{Col, Filter, Table};
-use widgetry::{ControlState, EventCtx, Filler, Line, Panel, State, Text, Toggle, Widget};
+use widgetry::{
+    EventCtx, Filler, GfxCtx, Line, Outcome, Panel, State, TabController, Text, Toggle, Widget,
+};
 
+use super::generic_trip_table::{open_trip_transition, preview_trip};
 use crate::app::App;
 use crate::common::{checkbox_per_mode, cmp_duration_shorter, color_for_mode};
-use crate::sandbox::dashboards::generic_trip_table::GenericTripTable;
 use crate::sandbox::dashboards::DashTab;
 
-pub struct FinishedTripTable;
-
-impl FinishedTripTable {
-    pub fn new(ctx: &mut EventCtx, app: &App) -> Box<dyn State<App>> {
-        GenericTripTable::new(
-            ctx,
-            app,
-            DashTab::FinishedTripTable,
-            make_table_finished_trips(app),
-            make_panel_finished_trips,
-        )
-    }
+pub type Transition = widgetry::Transition<App>;
+
+pub struct TripTable {
+    tab: DashTab,
+    table_tabs: TabController,
+    panel: Panel,
+    finished_trips_table: Table<App, FinishedTrip, Filters>,
+    cancelled_trips_table: Table<App, CancelledTrip, Filters>,
+    unfinished_trips_table: Table<App, UnfinishedTrip, Filters>,
 }
 
-pub struct CancelledTripTable;
-
-impl CancelledTripTable {
-    pub fn new(ctx: &mut EventCtx, app: &App) -> Box<dyn State<App>> {
-        GenericTripTable::new(
-            ctx,
-            app,
-            DashTab::CancelledTripTable,
-            make_table_cancelled_trips(app),
-            make_panel_cancelled_trips,
-        )
+impl TripTable {
+    pub fn new(ctx: &mut EventCtx, app: &App) -> Self {
+        let mut tabs = TabController::new("trips_tabs");
+
+        let (finished, unfinished) = app.primary.sim.num_trips();
+        let mut cancelled = 0;
+        // TODO Can we avoid iterating through this again?
+        for (_, _, _, maybe_dt) in &app.primary.sim.get_analytics().finished_trips {
+            if maybe_dt.is_none() {
+                cancelled += 1;
+            }
+        }
+        let total = finished + cancelled + unfinished;
+
+        let finished_trips_btn = ctx
+            .style()
+            .btn_tab
+            .text(format!(
+                "{} ({:.1}%) Finished Trips",
+                prettyprint_usize(finished),
+                if total > 0 {
```

Comment:
Could I suggest refactoring a little helper here:

```
fn percent(x: usize, total: usize) -> f64 {
  if total > 0 {
    (x as f64) / (total as f64) * 100.0
  } else {
    0.0
  }
}
```

I halfheartedly started https://github.com/a-b-street/abstreet/blob/master/geom/src/percent.rs a while back in the hopes of cleaning up stuff like this; maybe it's worth pursuing it further later. But for this PR, at least a little helper function would be nice!

Reply:
ha ok, but for the record this is just your code cut/pasted. 😛

Reply:
This is a classic code review technique -- make the bar unrealistically high for the next unfortunate soul who touches old code, especially when the reviewer is the original author. Sorry to subject you to this! :D

Reply:
Done, and I also re-ordered the text. It's less "literate" but since these are effectively headers, I think it makes sense to have the static part first.

<img width="521" alt="Screen Shot 2021-03-19 at 11 25 38 AM" src="https://user-images.githubusercontent.com/217057/111826377-da658600-88a5-11eb-972d-9b9b1dbda5bf.png">